include/eve/module/core/regular/scan.hpp

```diff
@@ -7,49 +7,79 @@
 //==================================================================================================
 #pragma once
 
+#include <eve/arch.hpp>
 #include <eve/detail/overload.hpp>
+#include <eve/concept/invocable.hpp>
+#include <eve/concept/substitute.hpp>
 
 namespace eve
 {
-  //TODO DOC
-//================================================================================================
-//! @addtogroup core_simd
-//! @{
-//!   @var scan
-//!   @brief Computes the TODO
-//!
-//!   @groupheader{Header file}
-//!
-//!   @code
-//!   #include <eve/module/core.hpp>
-//!   @endcode
-//!
-//!   @groupheader{Callable Signatures}
-//!
-//!   @code
-//!   namespace eve
-//!   {
-//!      TODO
-//!   }
-//!   @endcode
-//!
-//!   **Parameters**
-//!
-//!      * `x`:  An instance of an [SIMD value](@ref eve::simd_value)
-//!
-//!    **Return value**
-//!
-//!      * TODO
-//!
-//!  @groupheader{Example}
-//!
-//!  TODO
-//! @}
-//================================================================================================
-EVE_MAKE_CALLABLE(scan_, scan);
+  template<typename Options>
+  struct scan_t : callable<scan_t, Options>
+  {
+    template<simd_value Wide, eve::monoid<Wide> Op, eve::substitute_for<Wide> Zero>
+    constexpr EVE_FORCEINLINE Wide operator()(Wide w, Op op, Zero z) const noexcept
+      requires (requires { as_value(z, as<Wide>{}); })
+    {
+      return EVE_DISPATCH_CALL(w, op, z);
+    }
+
+    template<simd_value Wide>
+    constexpr EVE_FORCEINLINE Wide operator()(Wide w) const noexcept
+    {
+      return EVE_DISPATCH_CALL(w, add, eve::zero);
+    }
+
+    EVE_CALLABLE_OBJECT(scan_t, scan_);
+  };
+
+  //================================================================================================
+  //! @addtogroup core_simd
+  //! @{
+  //!   @var scan
+  //!   @brief Computes the generalized prefix sum over a [simd value](@ref eve::simd_value).
+  //!
+  //!   @groupheader{Header file}
+  //!
+  //!   @code
+  //!   #include <eve/module/core.hpp>
+  //!   @endcode
+  //!
+  //!   @groupheader{Callable Signatures}
+  //!
+  //!   @code
+  //!   namespace eve
+  //!   {
+  //!     constexpr simd_value scan(simd_value auto x, auto op, auto zero) noexcept; // 1
+  //!     constexpr simd_value scan(simd_value auto x) noexcept;                     // 2
+  //!   }
+  //!   @endcode
+  //!
+  //!   **Parameters**
+  //!
+  //!      * `x`:  An instance of an [SIMD value](@ref eve::simd_value)
+  //!      * `op`: The binary operation to apply.
```

Comment:
probably want to say that `op` commutative and associative. Or something about a monoid.